src/util/rust_util.rs

```diff
@@ -6,3 +6,56 @@ pub const fn min_of_usize(a: usize, b: usize) -> usize {
         a
     }
 }
+
+use std::cell::UnsafeCell;
+use std::mem::MaybeUninit;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+
+/// InitializeOnce creates an uninitialized value that needs to be initialized later. InitializeOnce
+/// guarantees the value is only initialized once. This type is used to allow more efficient reads.
+/// Unlike the `lazy_static!` which checks whether the static is initialized
+/// in every read, InitializeOnce has no extra check for reads.
+pub struct InitializeOnce<T> {
+    v: UnsafeCell<MaybeUninit<T>>,
+    initialized: AtomicBool,
+}
+
+impl<T> InitializeOnce<T> {
+    pub const fn new() -> Self {
+        InitializeOnce {
+            v: UnsafeCell::new(MaybeUninit::uninit()),
+            initialized: AtomicBool::new(false),
+        }
+    }
+
+    /// Initialize the value. This should be called before ever using the struct.
+    pub fn initialize(&self, val: T) {
+        if self
+            .initialized
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .is_ok()
+        {
```

Comment:
If another thread calls `get_ref` after the `compare_exchange` but before the `write` below, the thread will think it is already initialized.

Reply:
I have changed the implementation of `InitializeOnce`. I think it fixes this problem. 

I am now using `Once` to guarantee the initialization only happens once and other threads will be blocked when one thread is doing the initialization. Also the initial value is now only created once.